### DISCUSSION TITLE
Is there an "AllCombinations" convenience method in Verify?

### DISCUSSION THREAD

**Discussion by emilybache** [VerifyTests/Verify#1320] [Q&A] [answered]
Hi,

I've previously used this "test all combinations" feature in Approvals.Java 

https://gist.github.com/emilybache/0a3386840bf4c495ddfc0ad35e038575

and in python approvals

https://gist.github.com/emilybache/93ff869d0946de862a6633d58dccd15f

The idea is that with only a very small amount of test code you generate a lot of input combinations and verify them all.

Is there an equivalent in Verify? I couldn't find it if so.

Regards,
Emily Bache

**Comment by SimonCropp** [comment, marked as answer]
currently no. although i am happy to consider a pull request that adds that feature

**Reply by emilybache** [reply]
Thanks for the prompt and accurate answer.

**Comment by emilybache** [comment]
In case anyone is interested this is a solution that uses Verify to solve GildedRose:

https://gist.github.com/emilybache/b7aef33a8fe449d47c0a194a9451bd3c

**Comment by SimonCropp** [comment]
@emilybache i released a beta with a new experimental api https://github.com/VerifyTests/Verify/blob/main/docs/combinations.md

thoughts?

i will flesh out the docs shortly

**Reply by emilybache** [reply]
Looks good!
- I'm not sure if you need the optional argument "captureExceptions", I'd expect it to do that in any case. 
- The ApprovalTests.Java implementation has an exception your method under test can throw "SkipCombination" if it wants to just ignore that one. I've not found it hugely important but might be worth including
- The ApprovalTests.Java implementation also has a "verifyBestCoveringPairs" method as an alternative to verifying all combinations. It cuts down the number of combinations considerably while preserving most of the variety. You might consider including it too - I think it's more practical to use in general.

**Reply by SimonCropp** [reply]
how should the output look if there a newlines in the exception message?

**Reply by emilybache** [reply]
I think you just include the newlines in the output too. Having said that - I wouldn't include a stacktrace, and particularly not line numbers that refer to the sourcecode. The approved file is supposed to be human-readable and doesn't need to strictly adhere to one line per combination. Of course it does make the line-base diff easier to read if it is succinct like that, but in practice I've found it's not a problem if a combination takes several lines. You can usually work out which lines belong to each combination.

**Reply by SimonCropp** [reply]
@emilybache where are the docs for verifyBestCoveringPairs  ?

**Reply by emilybache** [reply]
Good point! There probably isn't any. It's basically applying [All-Pairs testing](https://en.wikipedia.org/wiki/All-pairs_testing) to cut down the number of combinations.

**Reply by SimonCropp** [reply]
i pushed a stable V27.1.0 
updated the docs https://github.com/VerifyTests/Verify/blob/main/docs/combinations.md

**Reply by SimonCropp** [reply]
i doubt i will get around to implementing BestCoveringPairs

**Reply by emilybache** [reply]
That looks brilliant! You've spent a lot of effort on making the output look readable. Thanks for doing this.
